### ISSUE TITLE
Add an Option for automatic resolving of Enum-typed keys in Maps

### ISSUE THREAD

**Issue by mdindoffer** [victools/jsonschema-generator#277]
# Motivation
Consider a simple java map with Enum keys:
```java
Map<MyEnum, MyValueClass> myMap;
```
It should be possible to automatically generate a schema with a constraint on the string keys to only allow the defined `MyEnum` values.

# Current state

The `MAP_VALUES_AS_ADDITIONAL_PROPERTIES` deals only with Map values.

The [docs](https://victools.github.io/jsonschema-generator/#how-to-represent-a-map-lt-k-v-gt-in-a-generated-schema) mention only one "workaround" of using a custom regex pattern for the string key.
However, this is only a partial solution since:
- it requires constant tinkering each time the enum is modified
- the functionality is not exposed to the maven plugin options? (I might be wrong, but I didn't find anything)

# Proposed solution
Add a new standard `Option` for out-of-the-box flattening of enums to key constraints. Ideas on how to do this:
A) (preferred) Use `properties` [keyword](https://json-schema.org/understanding-json-schema/reference/object.html#id3) as described in this SO post https://stackoverflow.com/questions/39378941/jsonschema-validating-keys/39385260#39385260 
B) The generator should be able to convert the list of `MyEnum` values into a regular expression automatically and include that in the JsonSchema.

**Comment by CarstenWickner** [comment]
Hi @mdindoffer,

Thanks for your input.
Enums seem to be a constant source of questions regarding schema generation. That's why several built-in ways of representing an enum and the various types of serializing its values exist already.
Offering a built-in way for a map's enum keys would require to apply all those different kinds of serialization once more.

### TL;DR
I'll not offer this as built-in `Option` for now.
Instead, I'd encourage you to write a `CustomDefinitionProvider` that suits your needs, such as in the following examples, as requested.
In neither example is it required to touch this for enum values being added/removed and it can be made available to the Maven plugin.

### Option A
```java
public class EnumMapExampleA {

    public static void main(String[] args) throws JsonProcessingException {
        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON);
        configBuilder.with(new EnumMapModule());
        SchemaGenerator generator = new SchemaGenerator(configBuilder.build());
        JsonNode jsonSchema = generator.generateSchema(TestType.class);
        System.out.println(jsonSchema.toPrettyString());
    }

    static class TestType {
        public Map<MyEnum, TestType> map;
    }


    static enum MyEnum {
        FIRST_VALUE, SECOND_VALUE;
    }

    static class EnumMapModule implements Module {

        @Override
        public void applyToConfigBuilder(SchemaGeneratorConfigBuilder builder) {
            builder.forTypesInGeneral().withCustomDefinitionProvider(this::provideCustomSchemaDefinition);
        }

        private CustomDefinition provideCustomSchemaDefinition(ResolvedType javaType, SchemaGenerationContext context) {
            ResolvedType mapKeyType = context.getTypeContext().getTypeParameterFor(javaType, Map.class, 0);
            if (mapKeyType == null || !mapKeyType.getErasedType().isEnum()) {
                return null;
            }
            ResolvedType mapValueType = context.getTypeContext().getTypeParameterFor(javaType, Map.class, 1);
            ObjectNode schema = context.getGeneratorConfig().createObjectNode()
                    .put(context.getKeyword(SchemaKeyword.TAG_TYPE), context.getKeyword(SchemaKeyword.TAG_TYPE_OBJECT))
                    .put(context.getKeyword(SchemaKeyword.TAG_ADDITIONAL_PROPERTIES), false);                    ;
            ObjectNode propertiesNode = schema.with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES));
            EnumSet<? extends Enum> enumValues = EnumSet.allOf((Class) mapKeyType.getErasedType());
            enumValues.stream()
                    // map however the enum is being represented as a string value
                    .map(Enum::name)
                    .forEachOrdered(value -> propertiesNode.set(value, context.createStandardDefinitionReference(mapValueType, null)));
            return new CustomDefinition(schema, CustomDefinition.DefinitionType.STANDARD, CustomDefinition.AttributeInclusion.YES);
        }
    }
}
```
Running the above class produces this schema:
```json
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "map" : {
      "type" : "object",
      "additionalProperties" : false,
      "patternProperties" : {
        "(FIRST_VALUE)|(SECOND_VALUE)" : {
          "$ref" : "#"
        }
      }
    }
  }
}
```
### Option B

```java
public class EnumMapExampleB {

    public static void main(String[] args) throws JsonProcessingException {
        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON);
        configBuilder.with(new EnumMapModule());
        SchemaGenerator generator = new SchemaGenerator(configBuilder.build());
        JsonNode jsonSchema = generator.generateSchema(TestType.class);
        System.out.println(jsonSchema.toPrettyString());
    }

    static class TestType {
        public Map<MyEnum, TestType> map;
    }


    static enum MyEnum {
        FIRST_VALUE, SECOND_VALUE;
    }

    static class EnumMapModule implements Module {

        @Override
        public void applyToConfigBuilder(SchemaGeneratorConfigBuilder builder) {
            builder.forTypesInGeneral().withCustomDefinitionProvider(this::provideCustomSchemaDefinition);
        }

        private CustomDefinition provideCustomSchemaDefinition(ResolvedType javaType, SchemaGenerationContext context) {
            ResolvedType mapKeyType = context.getTypeContext().getTypeParameterFor(javaType, Map.class, 0);
            if (mapKeyType == null || !mapKeyType.getErasedType().isEnum()) {
                return null;
            }
            EnumSet<? extends Enum> enumValues = EnumSet.allOf((Class) mapKeyType.getErasedType());
            String regEx = enumValues.stream()
                    // map however the enum is being represented as a string value
                    .map(Enum::name)
                    .collect(Collectors.joining(")|(", "(", ")"));
            ResolvedType mapValueType = context.getTypeContext().getTypeParameterFor(javaType, Map.class, 1);
            ObjectNode schema = context.getGeneratorConfig().createObjectNode()
                    .put(context.getKeyword(SchemaKeyword.TAG_TYPE), context.getKeyword(SchemaKeyword.TAG_TYPE_OBJECT))
                    .put(context.getKeyword(SchemaKeyword.TAG_ADDITIONAL_PROPERTIES), false);                    ;
            schema.with(context.getKeyword(SchemaKeyword.TAG_PATTERN_PROPERTIES))
                    .set(regEx, context.createStandardDefinitionReference(mapValueType, null));
            return new CustomDefinition(schema, CustomDefinition.DefinitionType.STANDARD, CustomDefinition.AttributeInclusion.YES);
        }
    }
}
```
Running the above class produces this schema:
```json
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "map" : {
      "type" : "object",
      "additionalProperties" : false,
      "patternProperties" : {
        "(FIRST_VALUE)|(SECOND_VALUE)" : {
          "$ref" : "#"
        }
      }
    }
  }
}
```

### Summary

I personally prefer Option B for its brevity but you mentioned favoring Option A.
Either way, you may want to replace the `Enum::name` with whatever serialization logic you're applying to enums.

----

If you want to use such custom configurations, you need to wrap them into a `Module` and then make it available to the Maven plugin. In the  above example, I've already wrapped the configuration in the `EnumMapModule`, which you can then reference by fully qualified class name in your Maven plugin configuration, as per documentation:
https://victools.github.io/jsonschema-generator/#further-configurations-through-modules

**Comment by mdindoffer** [comment]
Ah, I apologize. This issue stems from my poor understanding of JsonSchema. Basically I hit a limitation of what it can describe, as it's meant to validate json objects, not define data structures of payloads. I thought it would be possible to generate a stable schema that would describe a `java.util.Map` collection with a predefined set of keys, but it looks like it isn't possible.

For a workaround, your suggestion with `additionalProperties` combined with a `patternProperties` regex is probably as close as it gets, although it's not what I was looking for, since no tool in the world can be expected to understand the regex and create a java enum out of that. I.e. if I'd try to translate the output schema from your examples back to Java World ™ it would look something like a Java top level object with a field of type "`map`" that has a `java.util.Map<String, Object>` field with the regex information stripped away.

Thanks for providing your solution, and I apologize again - we can close this.
